iocore/net/quic/test/test_QUICFrame.cc

```diff
@@ -1251,7 +1252,9 @@ TEST_CASE("QUICFrameFactory Create CONNECTION_CLOSE with a QUICConnectionError",
 
 TEST_CASE("QUICFrameFactory Create RST_STREAM with a QUICStreamError", "[quic]")
 {
-  QUICStream stream(new MockQUICRTTProvider(), new MockQUICConnection(), 0x1234, 0, 0);
+  auto mock_rtt        = std::make_unique<MockQUICRTTProvider>();
+  auto mock_connection = std::make_unique<MockQUICConnection>();
+  QUICStream stream(mock_rtt.get(), mock_connection.get(), 0x1234, 0, 0);
```

Comment:
I'm fine with this for now, but we should probably pass references instead of pointers. Since we have mock objects, we don't need to accept pointers (which implies we can accept nullptr but we don't check it in fact).

Reply:
They are still pointers :)

I was expecting something like this, but never mind. There is one place that require a pointer, maybe we can change it later though.
```diff
diff --git a/iocore/net/quic/QUICFlowController.cc b/iocore/net/quic/QUICFlowController.cc
index 57357f138..14896faf3 100644
--- a/iocore/net/quic/QUICFlowController.cc
+++ b/iocore/net/quic/QUICFlowController.cc
@@ -175,7 +175,7 @@ QUICLocalFlowController::update(QUICOffset offset)
 bool
 QUICLocalFlowController::_need_to_gen_frame()
 {
-  this->_threshold = this->_analyzer.expect_recv_bytes(2 * this->_rtt_provider->smoothed_rtt());
+  this->_threshold = this->_analyzer.expect_recv_bytes(2 * this->_rtt_provider.smoothed_rtt());
   if (this->_offset + this->_threshold > this->_limit) {
     return true;
   }
diff --git a/iocore/net/quic/QUICFlowController.h b/iocore/net/quic/QUICFlowController.h
index 8f9e3873f..516fca658 100644
--- a/iocore/net/quic/QUICFlowController.h
+++ b/iocore/net/quic/QUICFlowController.h
@@ -89,7 +89,7 @@ private:
 class QUICLocalFlowController : public QUICFlowController
 {
 public:
-  QUICLocalFlowController(QUICRTTProvider *rtt_provider, uint64_t initial_limit)
+  QUICLocalFlowController(QUICRTTProvider &rtt_provider, uint64_t initial_limit)
     : QUICFlowController(initial_limit), _rtt_provider(rtt_provider)
   {
   }
@@ -100,7 +100,7 @@ private:
   bool _need_to_gen_frame();
   QUICRateAnalyzer _analyzer;
 
-  QUICRTTProvider *_rtt_provider = nullptr;
+  QUICRTTProvider &_rtt_provider;
 };
 
 class QUICRemoteConnectionFlowController : public QUICRemoteFlowController
@@ -113,7 +113,7 @@ public:
 class QUICLocalConnectionFlowController : public QUICLocalFlowController
 {
 public:
-  QUICLocalConnectionFlowController(QUICRTTProvider *rtt_provider, uint64_t initial_limit)
+  QUICLocalConnectionFlowController(QUICRTTProvider &rtt_provider, uint64_t initial_limit)
     : QUICLocalFlowController(rtt_provider, initial_limit)
   {
   }
@@ -136,7 +136,7 @@ private:
 class QUICLocalStreamFlowController : public QUICLocalFlowController
 {
 public:
-  QUICLocalStreamFlowController(QUICRTTProvider *rtt_provider, uint64_t initial_limit, QUICStreamId stream_id)
+  QUICLocalStreamFlowController(QUICRTTProvider &rtt_provider, uint64_t initial_limit, QUICStreamId stream_id)
     : QUICLocalFlowController(rtt_provider, initial_limit), _stream_id(stream_id)
   {
   }
diff --git a/iocore/net/quic/QUICStream.cc b/iocore/net/quic/QUICStream.cc
index ff381473d..e7f71dd55 100644
--- a/iocore/net/quic/QUICStream.cc
+++ b/iocore/net/quic/QUICStream.cc
@@ -43,7 +43,7 @@
 static constexpr uint32_t MAX_STREAM_FRAME_OVERHEAD = 24;
 static constexpr uint32_t MAX_CRYPTO_FRAME_OVERHEAD = 16;
 
-QUICStream::QUICStream(QUICRTTProvider *rtt_provider, QUICConnectionInfoProvider *info, QUICStreamId sid,
+QUICStream::QUICStream(QUICRTTProvider &rtt_provider, QUICConnectionInfoProvider &info, QUICStreamId sid,
                        uint64_t recv_max_stream_data, uint64_t send_max_stream_data)
   : VConnection(nullptr),
     _info(info),
diff --git a/iocore/net/quic/QUICStream.h b/iocore/net/quic/QUICStream.h
index e70cbd906..892cd6b92 100644
--- a/iocore/net/quic/QUICStream.h
+++ b/iocore/net/quic/QUICStream.h
@@ -55,7 +55,7 @@ public:
       _state(nullptr, nullptr, nullptr, nullptr)
   {
   }
-  QUICStream(QUICRTTProvider *rtt_provider, QUICConnectionInfoProvider *info, QUICStreamId sid, uint64_t recv_max_stream_data = 0,
+  QUICStream(QUICRTTProvider &rtt_provider, QUICConnectionInfoProvider &info, QUICStreamId sid, uint64_t recv_max_stream_data = 0,
              uint64_t send_max_stream_data = 0);
   ~QUICStream();
   // void start();
@@ -102,7 +102,7 @@ protected:
   void _write_to_read_vio(QUICOffset offset, const uint8_t *data, uint64_t data_length, bool fin);
 
   QUICStreamErrorUPtr _reset_reason = nullptr;
-  QUICConnectionInfoProvider *_info = nullptr;
+  QUICConnectionInfoProvider &_info;
   QUICStreamId _id                  = 0;
   QUICOffset _send_offset           = 0;
 

```

Reply:
:-)